### PR TITLE
Upsteam DNS enhancement v2

### PR DESCRIPTION
**Pull request by ChadBHowell**
Rebased against latest devel branch per @dschaper

The purpose of these commits is to do the following:

1) Bring the dnsmasq.conf changes to be inline with how config files are modified on current Debian type systems by using the "*.d" directory. This allows for a cleaner distro upgrade path and avoids stomping our config changes during dnsmasq upgrades.

2) Allow the user to choose between Google and DynDNS upstream servers during the install. This will later be extended to allow the user to enter custom upstream DNS servers as well.

### CODE REVIEW

**Comment by dschaper** [comment]
Thanks for the PR. One thing I'd like to see is that on existing installs there is the `/etc/dnsmasq.conf` file that has the current configuration, and a copy of the original supplied config at `/etc/dnsmasq.conf.orig`. I'd like the script to check for the existence of those two files, and if they exist, move `/etc/dnsmasq.orig` back to `/etc/dnsmasq` and then drop the `01-pihole.conf` into the proper directory. This will help with preventing any breakage of existing installs. Is that something you could add to the process?

**Comment by ChadBHowell** [comment]
I thought about this earlier and a problem that I see is as follows:
1) user installs pihole for the first time and the original dnsmasq.conf moves to dnsmasq.conf.orig.
2) user updats pihole to the latest version and the first installs dnsmasq.conf overwrites dnsmasq.conf.orig and the original file is lost.

I think the best we can do is pop a warning saying a default conf was installed and mv the backup to dnsmasq.conf.upgrade.  This would preserve the original dnsmasq.conf.orig and give us a backup of the current file.  This would only happen when someone upgrades pihole the first time.  After that we would be using the 01-pihole.conf.

File: automated install/basic-install.sh [hunk 1]
```diff
@@ -90,7 +90,7 @@ backupLegacyPihole(){
 		echo "Original Pi-hole detected.  Initiating sub space transport"
 		$SUDO mkdir -p /etc/pihole/original/
 		$SUDO mv /etc/dnsmasq.d/adList.conf /etc/pihole/original/adList.conf.$(date "+%Y-%m-%d")
-		$SUDO mv /etc/dnsmasq.conf /etc/pihole/original/dnsmasq.conf.$(date "+%Y-%m-%d")
+		$SUDO mv /etc/dnsmasq.d/01-pihole.conf /etc/pihole/original/dnsmasq.d/01-pihole.conf.$(date "+%Y-%m-%d")
```
Comment (dschaper) [inline automated install/basic-install.sh]: This doesn't need to be modified. The backupLegacyPihole() only was to check for Pi-Hole v1, and won't be triggered unless someone is running very old code. This function is set for depreciation soon anyways.

File: automated install/basic-install.sh [hunk 2]
```diff
@@ -260,6 +260,41 @@ setStaticIPv4(){
...
+	esac
+}
+
++versionCheckDNSmasq(){
+	# Check if /etc/dnsmasq.conf is from pihole.  If so replace with an original and install new in .d directory
+	dnsFile="/etc/dnsmasq.conf"
+	dnsSearch="addn-hosts=/etc/pihole/gravity.list"
+	
+	if grep -q $dnsSearch $dnsFile; then
+		$SUDO mv /etc/dnsmasq.conf /etc/dnsmasq.conf.orig
```
Comment (dschaper) [inline automated install/basic-install.sh]: Here, if dnsmasq.conf is a Pi-Hole conf then there already will be a dnsmasq.conf.orig, so we don't want to nuke the file. How about a 3 way switch? Copy dnsmasq.conf to dnsmasq.conf.pihole, then copy the dnsmasq.conf.orig back to dnsmasq.conf...

File: automated install/basic-install.sh (hunk 1 above)
Comment (ChadBHowell) [inline automated install/basic-install.sh]: Ok.  I will revert it back.

Comment (dschaper) [inline automated install/basic-install.sh]: Thanks, it's kind of confusing and there probably should be comments in there to note what the function is really designed for...

**Comment by dschaper** [comment]
I didn't see the versionCheckDNSmasq() before I made those earlier comments. I blame the bad coffee water I'm drinking. (I'll fix the coffee issue tomorrow...) I made some line notes and we can hash those out in there...

**Comment by ChadBHowell** [comment]
How about something like this:

```bash
versionCheckDNSmasq(){
	# Check if /etc/dnsmasq.conf is from pihole.  If so replace with an original and install new in .d directory
	dnsFile1="/etc/dnsmasq.conf"
	dnsFile2="/etc/dnsmasq.conf"
	dnsSearch="addn-hosts=/etc/pihole/gravity.list"
	
	# Check dnsmasq.conf for pihole magic
	if grep -q $dnsSearch $dnsFile1; then
		# If true, Check dnsmasq.conf.orig for pihole magic
		if grep -q $dnsSearch $dnsFile1; then
			# If true, use advanced/dnsmasq.conf.original
			$SUDO mv /etc/dnsmasq.conf /etc/dnsmasq.conf.orig
			$SUDO cp /etc/.pihole/advanced/dnsmasq.conf.original /etc/dnsmasq.conf
		else
			# If false, mv original file back
			$SUDO mv /etc/dnsmasq.conf /etc/dnsmasq.conf.orig
	# If false, This is a fresh install
	fi
	$SUDO cp /etc/.pihole/advanced/01-pihole.conf /etc/dnsmasq.d/01-pihole.conf
	$SUDO sed -i "s/@INT@/$piholeInterface/" /etc/dnsmasq.d/01-pihole.conf
	$SUDO sed -i "s/@DNS1@/$piholeDNS1/" /etc/dnsmasq.d/01-pihole.conf
	$SUDO sed -i "s/@DNS2@/$piholeDNS2/" /etc/dnsmasq.d/01-pihole.conf
}
```

**Comment by dschaper** [comment]
Looks good, just make sure your inner `if` is `fi`d.

**Comment by ChadBHowell** [comment]
Good catch on the fi. This is me coding without coffee again...

**Comment by dschaper** [comment]
I just took it to be very well formatted pseudo-code ;)

**Comment by ChadBHowell** [comment]
Ok, that should be the last of the coffeeless mistakes... Done for the night, I will fix anything else tomorrow.

**Comment by dschaper** [comment]
https://github.com/pi-hole/pi-hole/tree/CBH/UpstreamDNS

**Comment by ChadBHowell** [comment]
Merged latest and fix slight merge error.

**Comment by dschaper** [comment]
Okay, I'll create a new alpha branch. Thanks again.

**Comment by ChadBHowell** [comment]
Updates transferred to the alpha branch #267

**Pull request closed without merging**